### PR TITLE
Implement SourceManager basics

### PR DESCRIPTION
**Pull request by mcastorina**
None of this code is used yet, but since it started growing in length I wanted to get some eyes on it before using it in the engine / adding more functionality.

### CODE REVIEW

File: pkg/engine/filesystem.go [hunk 1]
```diff
@@ -37,7 +45,6 @@ func (e *Engine) ScanFileSystem(ctx context.Context, c sources.FilesystemConfig)
 	if err != nil {
 		return errors.WrapPrefix(err, "could not init filesystem source", 0)
 	}
-	fileSystemSource.WithFilter(c.Filter)
```
Comment (rosecodym) [inline pkg/engine/filesystem.go:40]: Why did you remove this?

File: pkg/sources/source_manager.go [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+		return nil
+	})
+	return <-ch
+}
+
+// ScheduleRun blocks until a resource is available to run the source, then
+// asynchronously runs it. Error information is lost in this case.
+func (s *SourceManager) ScheduleRun(ctx context.Context, handle handle, conn *anypb.Any) error {
+	if err := ctx.Err(); err != nil {
+		return err
```
Comment (rosecodym) [inline pkg/sources/source_manager.go:107]: Why isn't this check done in `Run`?

File: pkg/sources/source_manager.go [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+	headlessJobCounter    int64
+	// Pool limiting the amount of concurrent sources running.
+	pool errgroup.Group
+	// Downstream chunks channel to be scanned.
+	outputChunks chan *Chunk
+}
+
+// sourceTemplate is an aggregate struct to hold the Source and its name.
+type sourceTemplate struct {
+	name   string
```
Comment (rosecodym) [inline pkg/sources/source_manager.go:37]: Are you expecting to use `name` in logs and the like? I don't see it used anywhere else (but I might just be missing it).

File: pkg/sources/source_manager.go [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+		outputChunks: outputChunks,
+	}
+	for _, opt := range opts {
+		opt(&man)
+	}
+	return &man
+}
+
+// Manage informs the SourceManager to track and manage a Source.
+func (s *SourceManager) Manage(ctx context.Context, name string, source Source) (handle, error) {
```
Comment (rosecodym) [inline pkg/sources/source_manager.go:72]: bleh, I am not wild about `Manage` as a verb here. What about `RegisterSource` or something else that alludes to the registration/ownership transfer that's happening?

File: pkg/sources/source_manager.go [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+	id, err := s.registerSource(ctx, name, source)
+	if err != nil {
+		return 0, err
+	}
+	s.handlesLock.Lock()
+	defer s.handlesLock.Unlock()
+	if _, ok := s.handles[id]; ok {
+		// TODO: smartly handle this?
+		return 0, fmt.Errorf("generated handle ID '%d' already in use locally", id)
```
Comment (rosecodym) [inline pkg/sources/source_manager.go:81]: I'm wondering if you could avoid having to worry about this weird case by establishing a facade that creates some sort of virtual API for headless managers.

File: pkg/engine/filesystem.go (hunk 1 above)
Comment (mcastorina) [reply pkg/engine/filesystem.go:40]: I moved it to right after the source is created (line 31). It changes the behavior by setting the filter before calling `Init`, but I checked the source and the order doesn't matter.

File: pkg/sources/source_manager.go (hunk 5 above)
Comment (mcastorina) [reply pkg/sources/source_manager.go:81]: Ah do you mean like making the `api` required and moving the `headlessJobCounter` etc. into an implementation of that?

I don't see how that would remove this edge case though so maybe I'm misunderstanding.

File: pkg/sources/source_manager.go (hunk 2 above)
Comment (mcastorina) [reply pkg/sources/source_manager.go:107]: Since `Run` is synchronous, the context cancellation should be caught downstream and returned as an error. Wouldn't hurt to add it there too though.

File: pkg/sources/source_manager.go (hunk 4 above)
Comment (mcastorina) [reply pkg/sources/source_manager.go]: Updated to `Enroll`

File: pkg/sources/source_manager.go [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+	id, err := s.api.RegisterSource(ctx, name, source.Type())
+	if err != nil {
+		return 0, err
+	}
+	handleID := handle(id)
+	s.handlesLock.Lock()
+	defer s.handlesLock.Unlock()
+	if _, ok := s.handles[handleID]; ok {
+		// TODO: smartly handle this?
+		return 0, fmt.Errorf("generated handle ID '%d' already in use locally", handleID)
```
Comment (rosecodym) [inline pkg/sources/source_manager.go:82]: is this error message still correct? (is the mention of use "locally" still the best way to describe the problem?)

**Review by rosecodym — APPROVED** [review]

**Review by rosecodym — APPROVED** [review]
this feels like it's really coming together, and i'm curious to see what it looks like as you flesh out more on the consumer side.

(as an aside: i can't tell whether some of your identifiers are still holdovers from previous implementations, so maybe give them a pass to see if there are any you'd like to update.)
